File: JavaScript/725-Split Linked List in Parts.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode[]}
 */
var splitListToParts = function(head, k) {
    let tmp = head, N = 0;
    
    while (tmp){
        tmp = tmp.next;
        N += 1;
    }
    
    let res = new Array();
    let runningHead = head;
    const first = N%k;
    const num = Math.floor(N/k);

    for (let i=0; i<k; i++){
        if (!runningHead){
            for (let x = i; x<k; x++){
                res.push(null);
            }
            return res;
        }
        res.push(runningHead);
        for (let j=0; j<num-1+(i<first); j++){
            runningHead = runningHead.next;
        }
        let prev = runningHead;
        runningHead = runningHead.next;
        prev.next = null;
    }
    
    return res;
};

File: C++/725-Split Linked List in Parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        vector<ListNode*> res;
        int num = 0;
        ListNode* curr = head;

        while (curr != nullptr) {
            num++;
            curr = curr->next;
        }

        int q = num/k, r = num%k;
        ListNode* node = head;
        for (int i=0; i<k; i++) {
            if (node == nullptr) {
                res.push_back(nullptr);
                continue;
            }

            ListNode* running = node;

            res.push_back(node);
            int cnt = 0;
            while (cnt != q + (r>0) - 1) {
                cnt++;
                running = running->next;
            }
            r--;

            node = running->next;
            running->next = nullptr;

        }

        return res;
    }
};